Src/HALAL/Services/FMAC/FMAC.cpp

```diff
@@ -0,0 +1,171 @@
+#ifdef FMAC_ERROR_CHECK
+	if(amount_to_preload > MemoryLayout.FInSize || amount_to_preload > MemoryLayout.FeedbackInSize){
+		ErrorHandler("Error while preloading data, cannot preload more data than the structure can hold");
+	}
+
+	while(HAL_FMAC_GetState(Instance.hfmac) != HAL_FMAC_STATE_READY){
+
+	}
+#endif
+
+	if (HAL_FMAC_FilterPreload(Instance.hfmac, preload_data, amount_to_preload, preload_feedback_data, amount_to_feedback_preload) != HAL_OK){
+		ErrorHandler("Unexpected error on preload of data");
+	}
+
+}
+
+void MultiplierAccelerator::software_run(int16_t* calculated_data, uint16_t* output_size){
+	Process.running_output_data = calculated_data;
+	Process.output_data_size = *output_size;
+	Process.state = MultiplierAccelerator::WAITING_DATA;
+	if(HAL_FMAC_FilterStart(Instance.hfmac, calculated_data, output_size) != HAL_OK){
+		ErrorHandler("Error while starting FMAC");
+	}
+}
+
+void MultiplierAccelerator::software_load_input(int16_t* input_data, uint16_t* input_size){
+	Process.input_data = input_data;
+	Process.input_data_size = *input_size;
+	if(HAL_FMAC_AppendFilterData(Instance.hfmac, input_data, input_size)  != HAL_OK){
+		ErrorHandler("Error while loading data into the FMAC");
+	}
+	Process.state = MultiplierAccelerator::RUNNING;
+}
+
+
+void MultiplierAccelerator::software_load_repeat(int16_t* input_data, uint16_t* input_size){
+	Process.state = MultiplierAccelerator::RUNNING;
+
+	Process.running_output_data = Process.output_data;
+	if(HAL_FMAC_ConfigFilterOutputBuffer(Instance.hfmac, Process.running_output_data, &Process.output_data_size) != HAL_OK){
+		ErrorHandler("Error while preparing buffer for the FMAC");
+	}
+
+	Process.input_data = input_data;
+	Process.input_data_size = *input_size;
+	if(HAL_FMAC_AppendFilterData(Instance.hfmac, input_data, input_size)  != HAL_OK){
+		ErrorHandler("Error while loading data into the FMAC");
+	}
+}
+
+void MultiplierAccelerator::software_load_full(int16_t* input_data, uint16_t* input_size, int16_t* output_data, uint16_t* output_size){
+	Process.state = MultiplierAccelerator::RUNNING;
+
+	Process.running_output_data = output_data;
+	Process.output_data_size = *output_size;
+	if(HAL_FMAC_ConfigFilterOutputBuffer(Instance.hfmac, output_data, output_size) != HAL_OK){
+		ErrorHandler("Error while preparing buffer for the FMAC");
+	}
+
+	Process.input_data = input_data;
+	Process.input_data_size = *input_size;
+	if(HAL_FMAC_AppendFilterData(Instance.hfmac, input_data, input_size)  != HAL_OK){
+		ErrorHandler("Error while loading data into the FMAC");
+	}
+}
+
+void MultiplierAccelerator::software_stop(){
+	if(HAL_FMAC_FilterStop(Instance.hfmac) != HAL_OK){
+		ErrorHandler("Error while stopping FMAC");
+	}
+}
+
+bool MultiplierAccelerator::is_ready(){
+	return Instance.hfmac->State == HAL_FMAC_STATE_READY && Process.state != MultiplierAccelerator::RUNNING;
+}
+
+void HAL_FMAC_HalfOutputDataReadyCallback(FMAC_HandleTypeDef *hfmac){}
```

Comment:
I guess this is marked as _weak in the HAL, so I assume it is just left blank on purpose in case it is needed in the future?

Reply:
yes, it is just a placeholder.